### PR TITLE
Kernel diagonal

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

**Comment by samir-nasibli** [comment]
@Rubtsowa please add related tests names in description, and skipped tests as well.

**Review by shssf — APPROVED** [review]

File: dpnp/dpnp_algo/dpnp_algo_indexing.pyx [hunk 1]
```diff
@@ -58,6 +58,7 @@ __all__ += [
 
 
 ctypedef void(*custom_indexing_2in_1out_func_ptr_t)(void *, void * , void * , size_t)
+ctypedef void(*custom_indexing_2in_1out_func_ptr_t_)(void * , void * , size_t, size_t * , size_t * , size_t)
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_indexing.pyx:61]:
```suggestion
ctypedef void(*custom_indexing_2in_1out_func_ptr_t_)(void * , void * , const size_t, size_t * , size_t * , const size_t)
```
corresponding to you function signature.

File: dpnp/backend/kernels/dpnp_krnl_indexing.cpp [hunk 2]
```diff
@@ -23,10 +23,129 @@
...
 
+#include <iostream>
+#include <list>
+#include <vector>
+
 #include <dpnp_iface.hpp>
 #include "dpnp_fptr.hpp"
 #include "queue_sycl.hpp"
 
+using namespace std;
```
[the remark this answers is not in the document]
Comment (shssf) [reply dpnp/backend/kernels/dpnp_krnl_indexing.cpp]: in cpp files it is common practice except some special cases.

File: dpnp/backend/include/dpnp_iface.hpp [hunk 3]
```diff
@@ -320,6 +320,20 @@ INP_DLLEXPORT void dpnp_cov_c(void* array1_in, void* result1, size_t nrows, size
...
 
+/**
+ * @ingroup BACKEND_API
+ * @brief math library implementation of diagonal function
+ *
+ * @param [in]  array   Input array with data.
+ * @param [out] result  Output array.
+ * @param [in]  offset  Offset of the diagonal from the main diagonal.
+ * @param [in]  shape   Shape of input array.
+ * @param [in]  shape   Shape of output array.
```
Comment (shssf) [inline dpnp/backend/include/dpnp_iface.hpp:331]: same name? :-)

File: dpnp/backend/kernels/dpnp_krnl_indexing.cpp [hunk 4]
```diff
@@ -23,10 +23,126 @@
...
+#include <list>
+#include <vector>
+
 #include <dpnp_iface.hpp>
 #include "dpnp_fptr.hpp"
 #include "queue_sycl.hpp"
 
+
+template <typename _DataType>
+class dpnp_diagonal_c_kernel;
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_indexing.cpp:36]: it looks like unused type

**Review by shssf — APPROVED** [review]

**Pull request merged**